src/async_impl/decoder.rs

```diff
@@ -0,0 +1,298 @@
+use std::fmt;
+use std::mem;
+use std::io::{self, Read};
+use std::marker::PhantomData;
+
+use libflate::non_blocking::gzip;
+use tokio_io::AsyncRead;
+use tokio_io::io as async_io;
+use futures::{Async, Future, Poll, Stream};
+use futures::stream::Concat2;
+use hyper::StatusCode;
+use serde::de::DeserializeOwned;
+use serde_json;
+use url::Url;
+
+use header::{Headers, ContentEncoding, ContentLength, Encoding, TransferEncoding};
+use super::{body, Body, Chunk};
+use error;
+
+/// A potentially asynchronous driver for a stream of chunks.
+pub trait PollChunk {
+    fn poll_chunk(&mut self) -> io::Result<Option<Chunk>>;
+}
+
+/// A `Read`able wrapper over a stream of chunks.
+struct ReadableBody<TBodyStream> {
+    state: ReadState,
+    stream: TBodyStream,
+}
+
+impl<TBodyStream> ReadableBody<TBodyStream> {
+    fn new(body: TBodyStream) -> Self {
+        ReadableBody {
+            state: ReadState::NotReady,
+            stream: body,
+        }
+    }
+}
+
+enum ReadState {
+    Ready(Chunk, usize),
+    NotReady,
+    Eof,
+}
+
+impl<TBodyStream> Read for ReadableBody<TBodyStream> 
+    where TBodyStream: PollChunk
+{
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        use std::cmp;
+
+        loop {
+            let ret;
+            match self.state {
+                ReadState::Ready(ref mut chunk, ref mut pos) => {
+                    let chunk_start = *pos;
+                    let len = cmp::min(buf.len(), chunk.len() - chunk_start);
+                    let chunk_end = chunk_start + len;
+                    buf[..len].copy_from_slice(&chunk[chunk_start..chunk_end]);
+                    *pos += len;
+                    if *pos == chunk.len() {
+                        ret = len;
+                    } else {
+                        return Ok(len);
+                    }
+                },
+                ReadState::NotReady => {
+                    match self.stream.poll_chunk() {
+                        Ok(Some(chunk)) => {
+                            self.state = ReadState::Ready(chunk, 0);
+                            continue;
+                        },
+                        Ok(None) => {
+                            self.state = ReadState::Eof;
+                            return Ok(0);
+                        },
+                        Err(e) => {
+                            return Err(e);
+                        }
+                    }
+                },
+                ReadState::Eof => return Ok(0),
+            }
+            self.state = ReadState::NotReady;
+            return Ok(ret);
+        }
+    }
+}
+
+/// A lazily constructed response decoder.
+pub struct LazyDecoder<TBodyStream>(LazyDecoderInner<TBodyStream>);
+
+enum LazyDecoderInner<TBodyStream> {
```

Comment:
Actually... Looking at this now I think this could all go away by just converting the gzip encoder into inner. So it doesn't matter if the body stream is already in a decoder.